### DISCUSSION TITLE
VSCode: how to combine Marp preview with Markdown-It contributions from other extensions?

### DISCUSSION THREAD

**Discussion by christianvoigt** [marp-team/marp#176] [Q&A] [answered]
Hi, I am the developer of [Argdown](https://argdown.org/). I am currently trying to figure out how to integrate Argdown with Marp.

The Argdown VSCode extension already contributes a [MarkdownIt plugin](https://www.npmjs.com/package/@argdown/markdown-it-plugin) to the Markdown preview to transform Argdown code snippets into argument maps. However, this does not work if I activate Marp for the Markdown document (using `marp: true`). Do you use VSCode's MarkdownIt instance at all? I think it would be great if this would work "out of the box".

If you are using your own instance of MarkdownIt, is it possible to configure it to install plugins? So far I have only found out that I can add Markdown plugins to the Marp cli.

**Comment by yhatt** [comment, marked as answer]
Thank you for your interest in contributing to Marp extension! 😆

> The Argdown VSCode extension already contributes a [MarkdownIt plugin](https://www.npmjs.com/package/@argdown/markdown-it-plugin) to the Markdown preview to transform Argdown code snippets into argument maps. However, this does not work if I activate Marp for the Markdown document (using `marp: true`). Do you use VSCode's MarkdownIt instance at all? I think it would be great if this would work "out of the box".

If marked the document as Marp Markdown via `marp: true`, Marp for VS Code will try to parse and render Markdown by using Marp's markdown-it engine instead of VS Code's markdown-it. markdown-it plugins for VS Code's engine provided by the VS Code extension are not covered in Marp engine.

[Pre-released versions](https://github.com/marp-team/marp-vscode/releases?after=v0.2.0) had actually supported extending Marp engine through other extensions. But we could not take this approach because there were feedbacks from early users about confliction with other major extensions, such as marp-team/marp-vscode#9. Because the most of extensions are designed for VS Code preview, not for Marp, we have thought that to keep extensibility as the same level as VS Code's markdown extensions will be hard in the point of view from maintainability and security[^security].

[^security]: https://github.com/marp-team/marp/discussions/164#discussioncomment-1313187

> If you are using your own instance of MarkdownIt, is it possible to configure it to install plugins? So far I have only found out that I can add Markdown plugins to the Marp cli.

Not yet. We had planned markdown-it plugin support powered by a project setting for Marp CLI in marp-team/marp-vscode#194, but I feel your suggestion looks sweet too. We are welcome to discuss about the design of contribution point for extensible Marp engine 😄

**Comment by christianvoigt** [comment]
Thanks for your answer, I should have looked  for related issues in marp-team/marp-vscode first.

I just did a little bit of research on how two extensions can communicate which each other. The most helpful information I could find was this [Stackoverflow answer](https://stackoverflow.com/questions/50058517/how-to-communicate-between-vscode-extensions) by Matt Bierner.

If I understand it correctly, I could call `extensions.extensions.getExtension('extension1.id')` with the Marp extension id and then use the `exports` property to get the return value of its activate function. If that would give me access to the Marp MarkdownIt engine, the Argdown extension could check if the Marp extension is installed and if that is the case add its MarkdownIt plugin to the engine. 

Edit: Here is [an example](https://code.visualstudio.com/api/references/vscode-api#extensions) in the docs.

I am not sure if that would always work from the activate function of the Argdown extension (maybe the Marp extension will be activated after the Argdown extension. Would that be a problem?).

It would be even better if this would work in a more declarative manner. In that case the Marp extension would add its own contribution point that other extensions can use in their `package.json`. But I don't think the Marp extension can filter extensions by their contributions. And I do not know if it would be possible or make sense to listen to all extension activation events to check if they return contributions to the Marp MarkdownIt engine. These are all questions we would probably have to ask the VSCode developers.
